jump/src/context.rs

```diff
@@ -49,111 +53,217 @@ fn path_to_str(path: &Path) -> Result<&str, String> {
         .map_err(|e| format!("{e}"))
 }
 
-pub struct Boot {
-    pub name: String,
-    pub description: Option<String>,
+#[derive(Debug)]
+pub(crate) enum FileEntry {
+    Skip(usize),
+    Install((File, PathBuf)),
+    ScieTote((File, Vec<(File, PathBuf)>)),
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub(crate) struct Binding {
+    target: PathBuf,
+    process: Process,
+}
+
+impl Binding {
+    pub(crate) fn execute(self) -> Result<(), String> {
+        atomic_path(&self.target.clone(), Target::File, |lock| {
```

Comment:
Right here we pass a closure that only gets executed if the file at the target path does not exist. If you read the implementation of atomic_path, that's done with a file lock.

Reply:
As a by the by, you may be aware atomic makes skis, boots and bindings.

Reply:
For more clarity, the current structure of the ~/.nce is:
```
~/.nce/
  <file hash directory>/
    <file>.lck
    <file - expanded if archive or directory, blob if neither>
  ...
  <lift manifest hash directory>
    lift.lck
    lift.json
    bindings/
    locks/
      <binding cmd name empty file>
      <binding cmd name empty file>.lck
```

It's the last `<binding cmd name empty file>{,.lck}` pair that is used as the gate for each binding that runs in the bindings sandbox above.

Reply:
I guess I should have named that directory `gates`. Some atomics threading the gates right there: https://external-preview.redd.it/VpSdsY8FC78KrqgHfW1eS0aDTgCEoIOwNwtF5oI3zAU.jpg?auto=webp&s=ecd159eb357406f68bcde49a859b61a4c8a851ae

Reply:
Ah. Awesome - turns out the piece of the puzzle I missed out on was the `<lift manifest hash directory>/` - I didn't realize that was also stored in there.

👍🏽

Reply:
Would you say/recommend that the boot bindings should be written as idempotent commands? 

Or, I guess I might be thinking "side-effect free" but that's not what I really mean either.

Would you suggest that boot bindings be a command where the 'work' is encapsulated entirely within the `~/.nce` folder(s)?

Trying to come up with something adequately silly here as an example...

I package `awesomeapp 1.0` and it has a sqlite database in `/var/lib/awesomeapp/db.sqlite` or something similar.

I come up with `awesomeapp 2.0` and in the boot binding, I have an awful command/script that does something like this (without handling OS errors):
```
1. Backup /var/lib/awesomeapp/db.sqlite to /var/lib/awesomeapp/db-{timestamp}.sqlite to 
2. Rename /var/lib/awesomeapp/db.sqlite to /var/lib/awesomeapp/db2.sqlite
3. Run DB migration on /var/lib/awesomeapp/db2.sqlite
```

This is a script that can/will fail in any number of ways, but with regards to scie specifically:

- Should this paradigm of command even be used (filesystem side-effects elsewhere on the OS)
- If we have a binding that will always fail (e.g. incorrect directory permissions to perform step 1), will the `awesomeapp 2.0` executable just never run, as it's pre-requisite binding can't be successfully completed (without me fixing directory permissions)?

Reply:
If the binding always fails, the app never runs. You'd need to fix permissions or intervene in some way.

As far as what to rule in / out, you should just be aware the lock is cooperative and so you just have to think about non-cooperative uses. If you try to mess with paths not in your cooperative control, like /usr/local, you may have to contend with a package manager running concurrently to you. I can't help you there!

This install bindings mechanism only guarantees:
1. Your binding runs (successfully) only once.
2. You have `*`complete control of `{scie.binding}` during that run - it's dedicated to your app alone by dint of your app manifest hash being unique.

`*` Obviously external commands could be malicious and guess your binding dir or else run blanket and catch you that way (`rm -rf ~/.nce`).

Reply:
It should be clear that in a container build step, you're solid. That environment has the guaranties of you running in isolation with nothing else concurrent. On a bare metal host, things are less solid. You probably want to stick to mutating the `scie.binding` dir. This will be the situation Pants is in installing itself sometimes on end user machines, sometimes in CI, sometimes in production.

Reply:
I've added https://github.com/a-scie/jump/issues/22 to the 0.2.x release and will work on that today. There needs to be a user guide obviously explaining how you build a scie including how you write the lift manifest / what features it supports.

Reply:
Great, thanks for the info! 

The "This install bindings mechanism only guarantees" snippet you added above would be great in the docs, as well as the bare-metal vs container difference. Those two snippets basically cleared up everything for me about binding operation.

Reply:
Ok, good to know. I'll add a section on boot bindings in a new PR on top of #24 after this and that land and be sure to include those bits.